### ISSUE TITLE
Lami 1.0 Spec precision on numbers

### ISSUE THREAD

**Issue by TheMatthew** [lttng/lami-spec#1]
The number object is specified as follows 

class: number
value 
low:
high:

where if value is specified is specified, low and high are optional, and if low AND high is specified, value is optional.

The examples below are taken directly from the specification.

Example 4. Number object without limits

{
  "class": "number",
  "value": -23.75
}

Example 5. Number object with limits

{
  "class": "number",
  "value": 19.72,
  "low": 15.99,
  "high": 23.18
}

Example 6. Number object without value

{
  "class": "number",
  "low": 20.52,
  "high": 24.18
}

Example 7. Number object with an infinite higher limit

{
  "class": "number",
  "low": 6675.92,
  "high": "+inf"
}

This looks like three data points are crammed into one definition. 
1. Points
2. Ranges
3. Ranges with points (example 5)

An issue arrises with Example 5: Number object with limits. 

This is effectively saying that the number has a precision. But the precision does not follow the rules of uncertainty figures. It does not make sense to accept 
Min: -2 val : 0.0423 max : 2

This becomes very important if you decide to aggregate values as the errors need to add up properly.

**Comment by PSRCode** [comment]
>  But the precision does not follow the rules of uncertainty figures

Significant figures ?

This concern seems to have been addressed by @eepp on the gerrit patchset for the support of LAMI 1.0 by TraceCompass [1]:

> Philippe Proulx
> Jun 14 1:22 PM
> I think this is going too far in regard to LAMI.
> 0.8 ± 1.0 is valid: the range is [-0.2, 1.8].
> AFAIK there's no notion of significant digits and scales in LAMI, so all numbers can be considered to have infinite precision. That is, 0.8 ± 0.101 is okay... the range is [0.699, 0.901]. There's no way to encode a precision in a JSON number anyway. 

[1] https://git.eclipse.org/r/#/c/74333/4/analysis/org.eclipse.tracecompass.analysis.lami.core/src/org/eclipse/tracecompass/internal/provisional/analysis/lami/core/types/LamiData.java@394

**Comment by TheMatthew** [comment]
yes, change of sentence halfway through. was it still clear? uncertainty ... to significant figures.

**Comment by compudj** [comment]
The [min, max] ranges are an alternative way to express uncertainty around a value, which is more flexible than just having a number of significant digits.

You can deduce the number of significant digits in the value based on the [min, max] interval. If the value has more digits than what the min-max interval defines as uncertainty, just don't care about those more precise digits. This could happen if, for instance, an analysis does a floating point division by 3, without removing the less significant digits, and end up with tons of useless numbers after the dot. Just use the min/max interval to deduce the number of significant digits if you care from a consumer perspective.

**Comment by TheMatthew** [comment]
> You can deduce the number of significant digits in the value based on the [min, max] interval. If the value has more digits than what the min-max interval defines as uncertainty, just don't care about those more precise digits. This could happen if, for instance, an analysis does a floating point division by 3, without removing the less significant digits, and end up with tons of useless numbers after the dot. Just use the min/max interval to deduce the number of significant digits if you care from a consumer perspective.

This means if I understand correctly, that there is an inherent uncertainty, which is a good thing. It means though that we need to take care of this on at the viewer level.

However, min/max/val still is a dangerous combo with no context. An example: the trace ends and you don't know when your interval finishes. IMO the good way to go about it would be to have an interval where min is ((trace end)-(interval start)) and the max is +inf. 
The less good way would be to have min and max the same as before and val as something in between. 

Basically it should be very clear when to use the triplet.

**Comment by eepp** [comment]
> Basically it should be very clear when to use the triplet.

Yes, but it's still the producer's problem. The consumer needs to be ready to receive and deal with anything. I should specify that `min` must be lesser than or equal to `max`, and that `value` must be in between, otherwise the number is considered invalid.

**Comment by compudj** [comment]
Latest example from Matthew implies a time range. A time range is represented with two time values, not only one. For instance, let's suppose we have an interrupt handler which starts in the middle of the trace, but ends after the end of trace. Its time range starts at a timestamp value in the middle of the trace (possibly with a mix/max for uncertainty), and ends at a timestamp for which the actual end value is unknown, but min=end of trace (possibly minus uncertainty on the end of trace timestamp), max=+inf.

I agree with Phil on specifying that min <= value <= max (if the values are there) else should be considered invalid.

**Comment by TheMatthew** [comment]
So to be clear, the responsibility of the viewer is to validate that min <= val <= max and that is all?

edit: The viewer only needs to validate m<v<m, that is all it needs to do.

**Comment by eepp** [comment]
The viewer needs to validate that:
- If `min` exists, then `max` exists also, and vice versa
- If `min` and `max` are absent, then `value` exists
- If `min` and `max` exists, then `min <= max` 
  - If `value` also exists, then `min <= value <= max`
- `min` and `max` are either JSON **numbers** or `-inf` for `min`, `+inf` for `max`
- `value` is a JSON **number**

The viewer should also use those properties in some way. In a table, it should display the boundaries textually, either in dedicated columns, or by formatting the cell accordingly, for example `2.24 ± 0.05`. On a chart, it should draw little error bars, or something alike depending on the type of chart. If `min` or `max` is the infinity, then in some contexts the user could be warned that some data were excluded because there's no way to know where they start or finish.
